### ISSUE TITLE
Also delete toots

### ISSUE THREAD

**Issue by BartG95** [kensanata/mastodon-backup#18]
It would be nice if there was an option `--retention-period` or something like that, so that someone could invoke the program with `--retention-period=4w` . That would archive all toots that are older than four weeks and then delete them from the instance. Of course, this should also automatically archive the attachments, before deleting the toot.

**Comment by kensanata** [comment]
Hm, sounds interesting. I'd say I want all the retention period on the server to be different from the retention period in my archive, e.g. 4w online and forever locally. I'd also prefer some stronger option name `--delete-older-than 4w`.

**Comment by BartG95** [comment]
> I'd say I want all the retention period on the server to be different from the retention period in my archive, e.g. 4w online and forever locally.

Yes, I think that it should only delete from the instance. Probably this shouldn't be done while archieving, but just as a seperate command. So we would have the following commands:

- Archive
- Text/Html
- Media
- Delete old

Then users can write their scripts with these tools. So for me, I would create a cron job that runs every week and execute this in the following order: Archive -> Media -> Delete old

**Comment by kensanata** [comment]
I like it.

**Comment by BartG95** [comment]
Also, to prevent deleting toots that are not yet archived, the delete tool should only operate on the toots that are already in the archive.

**Comment by kensanata** [comment]
Totally untested: 52a1df8. Do you feel like testing it? Perhaps start with this change:
```python
        if collection == 'statuses':
            mastodon.delete(status["id"]);
            # print ("Delete %s: %s" % (status["id"], status["created_at"]))
        elif collection == 'favourites':
            mastodon.unfavourite(status["reblog"]["id"])
            # print ("Unfavour %s: %s" % (status["id"], status["created_at"]))
```
to
```python
        if collection == 'statuses':
            # mastodon.delete(status["id"]);
            print ("Delete %s: %s" % (status["id"], status["created_at"]))
        elif collection == 'favourites':
            # mastodon.unfavourite(status["reblog"]["id"])
            print ("Unfavour %s: %s" % (status["id"], status["created_at"]))
```
Perhaps we should add a `--dry-run` option which does that.

**Comment by BartG95** [comment]
The dry run seems to work. The actual run doesn't work, because the mastodon api is not imported and constructed. To fix this, some code can be copy-pasted from the archive function, but that seems to be redundant. 

Also, I would suggest to make `--older-than` required. It's a destructive operation, so I think people should very explicitly opt in.

**Comment by kensanata** [comment]
This is proving to be trickier than expected. See 4cb7450 and the [documentation](https://github.com/kensanata/mastodon-backup#expiring-your-toots-and-favourites).

**Comment by kensanata** [comment]
Still missing: what happens when you run expiration again and again – will you keep trying to delete older statuses and unfavour older favourites? This causes useless traffic. We should probably have save sort of marker telling us that a toot in our archive has already been deleted.

**Comment by BartG95** [comment]
Suggestion: move deleted toots to a .deleted.json file?

Also, maybe the json archive should be splitted anyway into months or something like that? (To prevent the archive from getting to big.) So you would have 2017.01.json, 2017.02.json, etc. Then users can choose to delete a (range of) month(s). (Where months are relative, not absolute.)

As soon as such a monthly archive is deleted, it would be changed to 2017.01.deleted.json or moved to a folder named deleted.

The best thing of course would be if this is included in mastodon itself: a setting that sends your toots from last month (via email) and then delete them.

Maybe I am overthinking this.

**Comment by kensanata** [comment]
Hm. Perhaps, in order to close this issue it would be enough to mark toots as deleted in the archive and be done with it. Then we can open a new issue revolving around the problem of growing file size, splitting files, stuff like that.

**Comment by BartG95** [comment]
Agreed. This is getting to complex for just one issue.

**Comment by kensanata** [comment]
You are welcome. Marking toots as deleted and not deleting them again is part of 7b2801b. Let me know how it goes and if it works, we'll close this issue. While testing it myself I ran into several small problems:
- failing to save for some reason or another left a size 0 file and the entire archive lost (now I'm renaming the archive before saving it)
- an archive with no content caused archiving to fail (couldn't parse the JSON, obviously)
If only there was a way to write tests for this kind of application – a Mastodon instance with no DB and no UI, just the API! If only I had gotten further with my [Gomphoterium](https://github.com/kensanata/gomphotherium) implementation (currently abandoned).

**Comment by BartG95** [comment]
````
PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 bartg95@mastodon.host
This is a dry run and nothing will be expired.
Instead, we'll just list what would have happened.
Use --confirmed to actually do it.
Traceback (most recent call last):
  File "C:\Program Files\Python36\Scripts\mastodon-archive-script.py", line 11, in <module>
    load_entry_point('mastodon-archive==0.0.3', 'console_scripts', 'mastodon-archive')()
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\__init__.py", line 108, in main
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\expire.py", line 62, in expire
TypeError: 'NoneType' object is not subscriptable

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe archive bartg95@mastodon.host
Get user info
Get statuses (this may take a while)
Get favourites (this may take a while)
Saving 379 statuses and 1 favourites
PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe media bartg95@mastodon.host
202 urls in your backup (half of them are previews)
Downloading |##################              | 115/202
Not Found: https://mastodon.host/media_proxy/2320039/small
Downloading |##################              | 116/202
Not Found: https://mastodon.host/media_proxy/2320039/original
Downloading |################################| 202/202
2 downloads failed

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 bartg95@mastodon.host
Loading existing archive
This is a dry run and nothing will be expired.
Instead, we'll just list what would have happened.
Use --confirmed to actually do it.
Delete: 2017-11-06 "[...]"
.
.
.
Delete: 2017-10-28 "[...]"

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 --confirmed bartg95@mastodon.host
Loading existing archive
Expiring |#                               | 1/18
We need to authorize the app to make changes to your account.
Traceback (most recent call last):
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\expire.py", line 76, in expire
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\expire.py", line 29, in delete
  File "C:\Program Files\Python36\lib\site-packages\mastodon\Mastodon.py", line 603, in status_delete
    return self.__api_request('DELETE', url)
  File "C:\Program Files\Python36\lib\site-packages\mastodon\Mastodon.py", line 1149, in __api_request
    raise MastodonAPIError("Mastodon API returned error: " + str(response['error']))
mastodon.Mastodon.MastodonAPIError: Mastodon API returned error: This action is outside the authorized scopes

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "C:\Program Files\Python36\Scripts\mastodon-archive-script.py", line 11, in <module>
    load_entry_point('mastodon-archive==0.0.3', 'console_scripts', 'mastodon-archive')()
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\__init__.py", line 108, in main
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\expire.py", line 80, in expire
  File "C:\Program Files\Python36\lib\site-packages\mastodon_archive-0.0.3-py3.6.egg\mastodon_archive\core.py", line 44, in deauthorize
NameError: name 'client_secret' is not defined
PS C:\Users\bartg\Downloads\Temp\masto>
````

**Comment by kensanata** [comment]
Ouch, this hurts.

What was the problem in the first run? `TypeError: 'NoneType' object is not subscriptable` makes me think you loaded an empty file or didn't have an archive, yet. Yep, I forgot to make the archive required. Fixed in a2e2d50.

Archiving and media download is OK. The second run was a dry run and that looks OK, too.

Now we get to the confirmed run. `We need to authorize the app to make changes to your account.` is correct. For archiving, we just need read permissions but for expiry we need write permissions. The problem is that deauthorizing the app runs into an error. I forgot to retest this after a refactoring. Gaah. Fixed in a9d5640.

**Comment by BartG95** [comment]
````
PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 bartg95@mastodon.host
You need to create an archive, first

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe archive bartg95@mastodon.host
Get user info
Get statuses (this may take a while)
Get favourites (this may take a while)
Saving 400 statuses and 1 favourites

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe media bartg95@mastodon.host
210 urls in your backup (half of them are previews)
Downloading |##################              | 123/210
Not Found: https://mastodon.host/media_proxy/2320039/small
Downloading |##################              | 124/210
Not Found: https://mastodon.host/media_proxy/2320039/original
Downloading |################################| 210/210
2 downloads failed

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 bartg95@mastodon.host
Loading existing archive
This is a dry run and nothing will be expired.
Instead, we'll just list what would have happened.
Use --confirmed to actually do it.
Delete: 2017-11-06 "[..]"
.
.
.
Delete: 2017-10-28 "[..]"

PS C:\Users\bartg\Downloads\Temp\masto> mastodon-archive.exe expire --older-than 5 --confirmed bartg95@mastodon.host
Loading existing archive
Expiring |#                               | 1/18
We need to authorize the app to make changes to your account.
Log in
Visit the following URL and authorize the app:
https://mastodon.host/oauth/authorize?client_id=xxxxxxxxxxxxxxxxxxxxxxxxxx&response_type=code&redirect_uri=urn%3Aietf%3Awg%3Aoauth%3A2.0%3Aoob&scope=read+write
Then paste the access token here:
xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Expiring |################################| 18/18

PS C:\Users\bartg\Downloads\Temp\masto>
````
Have a Gargy award (https://mastodon.social/@Gargy):
![](https://files.mastodon.social/media_attachments/files/002/214/227/original/4f73b022ec999d29.png)
Gargy award for making an awesome toot archiver.
